### PR TITLE
Runtime configs for tests

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR enables setting configuration for the integration tests at runtime.

In particular, it
* adds routing of the genesis proof, genesis ledger and runtime config for integration tests
* replaces `Genesis_constants` with `Runtime_config`
  - the intention is that all previously-compile-time constants will be moved here
* replaces `Precomputed_values` with a search for existing genesis proofs
  - currently, this falls back on generating them if none are found
  - eventually, we should tweak the CI test runner to generate them and upload them to the S3 bucket, so that in the future CI can avoid re-building the equivalent proofs
* replaces compile-time genesis proof generation in `Precomputed_values` with a lookup at runtime, falling back to generation if none is found
  - we will probably want to change this default for testnet releases
  - we can do this once on CI when running the tests and re-use the proof by storing it in the keys S3 bucket
* replaces `Test_genesis_ledger` with `Genesis_ledger`
  - all 'named' ledgers are (lazily) available at runtime
  - custom ledgers can be swapped in at runtime
  - a simpler, balances-only interface replaces most of the named ledger definitions
* removes `Genesis_ledger_helper`, and replaces `runtime_genesis_ledger.exe` with a coda subcommand `coda.exe generate-genesis-proof`
* adds arguments to the command line for the daemon and integration tests to specify/tweak the runtime configuration

Checklist:

- [ ] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

### CODE REVIEW

**Comment by mrmr1993** [comment]
> Actually, if the constants are changed at runtime, we might want to generate new proof based on those?

This should be the behaviour here already. If it isn't, that's a bug.

For reference, I'm issuing this as a series of smaller PRs to make it easier to review.

**Comment by deepthiskumar** [comment]
> > Actually, if the constants are changed at runtime, we might want to generate new proof based on those?
> 
> This should be the behaviour here already. If it isn't, that's a bug.

Possible that I misunderstood the changes. I'm looking at the flow in `coda.ml` where one can specify the constants using the daemon flags and load the genesis proof from a directory. This directory also contains `runtime_config.json` from where the constants for the genesis protocol state is obtained. These two set of constants could be different

**Comment by mrmr1993** [comment]
> This directory also contains `runtime_config.json` from where the constants for the genesis protocol state is obtained. These two set of constants could be different

The intention here was to mirror the behaviour of `runtime_genesis_ledger.exe`, where the configuration and the proof are associated with **the configuration that it overrides**, not the desired configuration. Personally, I think this behaviour is ill-advised, but I left it in in case we were depending on it. I'll gladly remove it if we don't need it.

If this is used as intended (ie. the `runtime_config.json` isn't manually modified), it should exactly match the config that is searched for; if we want the override behaviour, copying the desired configuration's directory over the overridden config's directory should also allow that.

**Comment by deepthiskumar** [comment]
The behavior before was:
1. Using `runtime_genesis_ledger.exe`, generate the genesis ledger and the genesis proof by optionally overriding the compile-time constants. Here, the constants that can be overridden are `k`, `delta`, `genesis_timestamp`, and `txpool_size`. Coda daemon then loads the proof, the ledger, and the constants
2. Coda daemon could optionally override the constants `delta`, `genesis_timestamp`, and `txpool_size` (Specified by`Genesis_constants.Daemon_config`) because these don't affect the genesis proof (#4781 removes `delta` and `genesis_timestamp` from protocol state to make this work). The  constants that affect the genesis proof are `k`, `c`, `block_window_duration`  which cannot be overridden by the daemon. Right now `c` and `block_window_duration` are required to be compile-time because they are needed for the key generation which happens at compile-time. 

But in this, PR `k` also can be overridden and if it does get overridden the genesis proof would become invalid.
Also, this PR adds the possibility of generating genesis proof as part of the initialization (using the flag `--generate-genesis-proof`)and therefore can allow `k` to be overridden only for that specific case. 

The idea originally was  to override constants that affect the proof only when generating the proof i.e., using `runtime_genesis_ledger.exe` and not having to rebuild the binary when they change.  Any other constant that doesn't affect the genesis ledger or the proof can be overridden by the daemon at runtime

**Comment by mrmr1993** [comment]
> But in this, PR `k` also can be overridden and if it does get overridden the genesis proof would become invalid.
> Also, this PR adds the possibility of generating genesis proof as part of the initialization (using the flag `--generate-genesis-proof`)and therefore can allow `k` to be overridden only for that specific case.

Indeed. In this case, the base hash will change, and so the proof for *that* base hash will be looked up instead, or generated if unavailable.

This is similarly true for accounts: changing the ledger will again change the base hash, triggering a find/generate for that new hash.

I'll change the behaviour to
* use the runtime configuration rather than the one bundled with the proof
* verify the genesis proof against our anticipated base hash before progressing

**Comment by deepthiskumar** [comment]
> Indeed. In this case, the base hash will change, and so the proof for _that_ base hash will be looked up instead, or generated if unavailable.
> 
> This is similarly true for accounts: changing the ledger will again change the base hash, triggering a find/generate for that new hash.
> 
> I'll change the behaviour to
> 
>     * use the runtime configuration rather than the one bundled with the proof
> 
>     * verify the genesis proof against our anticipated base hash before progressing

I think we still want to use the constants from the directory when `genesis-ledger-dir` is set. This is assuming the user specifically wanted to use the proof generated for the constants in the file.
if `genesis-ledger-dir` is not set then we look for it in other directories and expect it to match the constants? 
maybe @yourbuddyconner  has some opinions since this affects deployment  the most

**Comment by yourbuddyconner** [comment]
Overall I like this API, here are some things to consider on my end: 
- I like the approach of incorporating the genesis ledger feature-set into the core CLI
- I would like this API to be as close as possible to the Ethereum CLI, in that, when you pass the same chunk of Genesis JSON to each daemon on a private network, they will be compatible with each other
- I think this same logic should extend to our genesis constants, and we should gate the feature-set so end-users are warned away from it during public releases

File: Makefile [hunk 1]
```diff
@@ -67,20 +67,11 @@ GENESIS_DIR := $(TMPDIR)/coda_cache_dir
 
 # generate the actual ledger and store in a tar
 genesis_tar:
```
Comment (psteckler) [inline Makefile:69]: should `build` be a prerequisite for the build target?

File: src/app/cli/src/coda.ml [hunk 2]
```diff
@@ -912,7 +944,30 @@ let coda_commands logger =
...
   ; (Snark_flame_graphs.name, Snark_flame_graphs.command)
-  ; ("transaction-snark-profiler", Transaction_snark_profiler.command) ]
+  ; ("transaction-snark-profiler", Transaction_snark_profiler.command)
+  ; ( "generate-genesis-proof"
+    , Command.async
+        ~summary:"Generate a genesis proof for a runtime configuration"
+        (let open Command.Param in
+        let open Command.Let_syntax in
+        let%map tar_file =
+          flag "--tar-file" ~doc:"PATH path to the .tar.gz file to generate"
```
[the remark this answers is not in the document]
Comment (psteckler) [reply src/app/cli/src/coda.ml:954]: Note that #4702 puts the proof outside a `tar.gz` file. 

There will be conflicts with that PR, inevitably.

**Comment by mrmr1993** [comment]
Stale, closing.

**Pull request closed without merging**